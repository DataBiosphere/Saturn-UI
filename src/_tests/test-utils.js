import expect from 'expect-puppeteer'
import puppeteer from 'puppeteer'

const appUrlBase = 'http://localhost:3000/#workspaces'
const rawlsUrl = 'https://rawls.dsde-dev.broadinstitute.org'
//TODO: read this in from CircleCI

export async function getAccessToken() {
  const { auth } = require('google-auth-library')
  console.log('key is: ' + process.env['FIRECLOUD_DEV_SA_KEY_JSON'])
  const keys = JSON.parse(process.env['FIRECLOUD_DEV_SA_KEY_JSON'])


  const client = auth.fromJSON(keys)
  client.scopes = ['profile', 'email', 'openid', 'https://www.googleapis.com/auth/devstorage.full_control', 'https://www.googleapis.com/auth/cloud-platform']
  client.subject = 'emma.redwalker@test.firecloud.org'
  const token = await client.authorize().then(auth => auth.access_token)
  console.log('token is: ' + token)
  return token
}


export async function wait(ms) {
  return new Promise(resolve => setTimeout(resolve, ms))
}

export function generateUUID() {
  let d = new Date().getTime()
  const uuid = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, c => {
    const r = (d + Math.random()*16)%16 | 0
    d = Math.floor(d/16)
    return (c=='x' ? r : (r&0x3|0x8)).toString(16)
  })
  return uuid
}

export const setupTest = async (browser, page) => {
  puppeteer.DEFAULT_TIMEOUT_INTERVAL = 20001
  jest.setTimeout(20002)
  browser = await puppeteer.launch({
    args: ['--disable-features=site-per-process'], // this is needed to use waitForSelector within iframes
    headless: false//,
    // slowMo: 100 // try below option..
  })
  // TODO: add debug mode
  // if (DEBUG) {
  //   puppeteer.slowMo = 100 // slow down by X ms
  // }
  const context = await browser.createIncognitoBrowserContext()
  page = await context.newPage()
  await page.setViewport({
    width: 1920,
    height: 1200
  })
  page.on('console', msg => console.log('PAGE LOG:', msg.text()))

  await Promise.all([
    page.goto(appUrlBase, { waitUntil: 'networkidle0' })//,
  ])

  const token = await getAccessToken()
  console.log('token is: ' + token)

  // background signin
  const executionContext = await page.mainFrame().executionContext()
  await executionContext.evaluate(
    token => {
      window.forceSignIn(`${token}`)
    }, token
  )
  await page.waitForNavigation({ waitUntil: 'networkidle0' })

  // accept TOS on new environments
  try {
    await page.waitForSelector('[datatestid="acceptTosButton"]', { timeout: 2000 })
    await expect(page).toClick('[datatestid="acceptTosButton"]')
    await page.waitForNavigation({ waitUntil: 'networkidle0' })
  } catch (error) {
    console.log('Did not see the TOS page.. Continuing on.')
  }

  return [browser, page, token]
}

export const cleanupTest = async (browser, page, token, workspaceName, billingProjectName) => {
  console.log('post-test ðŸŽ‰')
  // TODO: if test failed, change the name of the screenshot
  const screenshotPath = 'screenshots/endoftest-batch-'+workspaceName+'-.png'
  console.log('screenshot saved at: ' + screenshotPath)
  await page.screenshot({ path: screenshotPath })
  // if debug, wait 1 second
  await wait(2000)
  // clean up workspace via api call
  await fetch(rawlsUrl + '/api/workspaces/' + billingProjectName + '/' + workspaceName,
    { 'headers': { 'authorization': 'Bearer ' + token, 'x-app-id': 'Saturn' }, 'body': null, 'method': 'DELETE', 'mode': 'cors' })
  // clean up in beforeEach as well?
  console.log('cleanup ðŸŽ‰ðŸŽ‰ðŸŽ‰')
  await page.close()
  await browser.close()
}

export const createWorkspace = async (page, workspaceName, billingProjectName) => {
  // verify that we're on the workspaces page
  await expect(page).toMatch('New Workspace')
  await wait(4000)

  await page.hover('[datatestid="createNewWorkspace"]')
  await page.click('[datatestid="createNewWorkspace"] div') // searches for a div descendant - the click does not launch the modal otherwise. TODO: investigate

  // create a new function to handle this?
  // note: all modals go inside: id="modal-root"
  await expect(page).toMatch('Workspace name *')
  await wait(4000)
  await page.type('[datatestid="workspaceNameInput"]', workspaceName)
  await page.click('[aria-label="billingProjectSelect"]')
  await page.type('[aria-label="billingProjectSelect"]', billingProjectName+'\n')

  await page.type('[placeholder="Enter a description"]', 'description for ' + workspaceName)
  await wait(2 * 1000)
  await expect(page).toClick('[datatestid="createWorkspaceButton"]')

  await page.waitForNavigation({ waitUntil: 'networkidle0' })
  await wait(2 * 1000) // try this and see if the workspace loading issue persists

  // confirm workspace created successfully
  await expect(page).toMatchElement('[datatestid="workspaceInfoSidePanel"]')
}

// waitForNetworkIdle is useful when you expect a network request without a page navigation
export async function waitForNetworkIdle(page, timeout, maxInflightRequests = 0) {
  page.on('request', onRequestStarted)
  page.on('requestfinished', onRequestFinished)
  page.on('requestfailed', onRequestFinished)

  let inflight = 0
  let fulfill
  const promise = new Promise(x => fulfill = x)
  let timeoutId = setTimeout(onTimeoutDone, timeout)
  return promise

  function onTimeoutDone() {
    page.removeListener('request', onRequestStarted)
    page.removeListener('requestfinished', onRequestFinished)
    page.removeListener('requestfailed', onRequestFinished)
    fulfill()
  }

  function onRequestStarted() {
    ++inflight
    if (inflight > maxInflightRequests) clearTimeout(timeoutId)
  }

  function onRequestFinished() {
    if (inflight === 0) return
    --inflight
    if (inflight === maxInflightRequests) timeoutId = setTimeout(onTimeoutDone, timeout)
  }
}
