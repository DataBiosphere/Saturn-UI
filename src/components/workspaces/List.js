import update from 'immutability-helper'
import Table from 'rc-table'
import { Component, Fragment } from 'react'
import { a, button, div, h, input, option, select } from 'react-hyperscript-helpers'
import _ from 'underscore'
import * as Ajax from '../../ajax'
import * as Nav from '../../nav'


class WorkspaceList extends Component {
  constructor(props) {
    super(props)
    this.state = {
      pageIndex: 1,
      itemsPerPage: 25,
      filter: '',
      listView: true,
      cardWidth: 5
    }
  }

  componentWillMount = () => {
    Ajax.rawls('workspaces').then(json =>
      this.setState({ workspaces: json })
    )
  }

  render() {
    if (this.state.workspaces) {
      const filteredWorkspaces = this.state.workspaces.filter(
        ({ workspace: { namespace, name } }) => `${namespace}/${name}`.includes(this.state.filter))
      const listPage = filteredWorkspaces.slice((this.state.pageIndex - 1) *
        this.state.itemsPerPage,
        this.state.pageIndex * this.state.itemsPerPage)

      const renderWorkspaceList = () => {
        return h(Table,
          {
            data: listPage,
            rowKey: ({ workspace }) => workspace.workspaceId,
            columns: [
              {
                title: 'Workspace', dataIndex: 'workspace', key: 'workspace',
                render: ({ namespace, name }) => {
                  return a({ href: Nav.getLink('workspace', namespace, name) },
                    `${namespace}/${name}`)
                }
              }
            ]
          }
        )
      }

      const renderWorkspaceCards = () => {
        return h(Fragment,
          [
            button(
              {
                onClick: () => this.setState(
                  prev => (update(prev, { cardWidth: { $apply: n => _.max([n - 1, 1]) } })))
              },
              '+'),
            button(
              {
                onClick: () => this.setState(
                  prev => (update(prev, { cardWidth: { $apply: n => n + 1 } })))
              },
              '-'),
            div({ style: { display: 'flex', flexWrap: 'wrap' } },
              _.map(listPage, ({ workspace: { namespace, name } }) => {
                return a({
                  style: {
                    display: 'block',
                    height: 100,
                    width: `${100 / this.state.cardWidth}%`,
                    border: '1px solid black',
                    boxSizing: 'border-box'
                  },
                  href: Nav.getLink('workspace', namespace, name)
                }, `${namespace}/${name}`)
              })
            )
          ]
        )
      }

      return h(Fragment, [
        input({
          placeholder: 'Filter',
          onChange: e => this.setState({ filter: e.target.value })
        }),
        button({ onClick: () => this.setState(prev => (update(prev, { $toggle: ['listView'] }))) },
          'Toggle display mode'),
        this.state.listView ? renderWorkspaceList() : renderWorkspaceCards(),
        div({ style: { marginTop: 10 } }, [
          'Page: ',
          select({
              style: { marginRight: '1rem' },
              onChange: e => this.setState({ pageIndex: parseInt(e.target.value, 10) }),
              value: this.state.pageIndex
            },
            _.map(_.range(1, filteredWorkspaces.length / this.state.itemsPerPage + 1),
              i => option({}, i))),
          'Items per page: ',
          select({
              onChange: e => this.setState({ itemsPerPage: parseInt(e.target.value, 10) }),
              value: this.state.itemsPerPage
            },
            _.map([10, 25, 50, 100],
              i => option({}, i)))
        ])
      ])

    } else {
      return 'Loading!'
    }
  }
}

const addNavPaths = () => {
  Nav.defRedirect({ regex: /^.{0}$/, makePath: () => 'workspaces' })
  Nav.defPath(
    'workspaces',
    {
      component: props => h(WorkspaceList, props),
      regex: /workspaces$/,
      makeProps: () => ({}),
      makePath: () => 'workspaces'
    }
  )
}

export { addNavPaths }
