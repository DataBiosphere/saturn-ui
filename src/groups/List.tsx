export const columnWidths = '1fr 30% 6rem 20px';
