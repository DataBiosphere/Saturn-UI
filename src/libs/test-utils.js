import * as Utils from 'src/libs/utils'


export const waitOneTickAndUpdate = wrapper => {
  return Utils.waitOneTick().then(() => wrapper.update())
}
