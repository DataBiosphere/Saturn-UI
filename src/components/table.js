import _ from 'lodash/fp'
import RCTable from 'rc-table'
import { Fragment, createRef } from 'react'
import { button, div, h, option, select, table, td, th, tr } from 'react-hyperscript-helpers'
import Interactive from 'react-interactive'
import Pagination from 'react-paginating'
import { Grid as RVGrid, ScrollSync as RVScrollSync } from 'react-virtualized'
import { icon } from 'src/components/icons'
import * as Style from 'src/libs/style'
import { Component } from 'src/libs/wrapped-components'


const paginatorButton = (props, label) => button(_.merge({
  style: {
    margin: '0 2px', padding: '0.25rem 0.5rem',
    border: '1px solid #ccc', borderRadius: 3,
    color: props.disabled ? Style.colors.disabled : Style.colors.primary,
    cursor: props.disabled ? 'not-allowed' : 'pointer'
  }
}, props), label)

/**
 * @param {number} props.filteredDataLength
 * @param {number} props.pageNumber
 * @param {function(number)} props.setPageNumber
 * @param {function(number)} [props.setItemsPerPage]
 * @param {number} props.itemsPerPage
 * @param {number[]} props.itemsPerPageOptions
 */
export const paginator = function(props) {
  const {
    filteredDataLength, pageNumber, setPageNumber, setItemsPerPage,
    itemsPerPage, itemsPerPageOptions = [10, 25, 50, 100]
  } = props

  return h(Pagination, {
    total: filteredDataLength,
    limit: itemsPerPage,
    pageCount: 5,
    currentPage: pageNumber
  }, [
    ({ pages, currentPage, hasNextPage, hasPreviousPage, previousPage, nextPage, totalPages, getPageItemProps }) => h(Fragment,
      [
        div({ style: { display: 'flex', justifyContent: 'flex-end', alignItems: 'center', marginRight: '1rem' } }, [
          `${(pageNumber - 1) * itemsPerPage + 1} - ${_.min([filteredDataLength, pageNumber * itemsPerPage])} of ${filteredDataLength}`,
          div({ style: { display: 'inline-flex', padding: '0.25rem 1rem' } }, [

            paginatorButton(
              _.merge({ disabled: currentPage === 1, style: { marginRight: '0.5rem' } },
                getPageItemProps({ pageValue: 1, onPageChange: setPageNumber })),
              [icon('angle-double left', { size: 12 })]
            ),

            paginatorButton(
              _.merge({ disabled: !hasPreviousPage, style: { marginRight: '1rem' } },
                getPageItemProps({ pageValue: previousPage, onPageChange: setPageNumber })),
              [icon('angle left', { size: 12 })]
            ),

            _.map(num => paginatorButton(
              _.merge({
                key: num,
                style: {
                  minWidth: '2rem',
                  backgroundColor: currentPage === num ? Style.colors.primary : undefined,
                  color: currentPage === num ? 'white' : Style.colors.primary,
                  border: currentPage === num ? Style.colors.primary : undefined
                }
              },
              getPageItemProps({ pageValue: num, onPageChange: setPageNumber })),
              num), pages
            ),

            paginatorButton(
              _.merge({ disabled: !hasNextPage, style: { marginLeft: '1rem' } },
                getPageItemProps({ pageValue: nextPage, onPageChange: setPageNumber })),
              [icon('angle right', { size: 12 })]
            ),

            paginatorButton(
              _.merge({ disabled: currentPage === totalPages, style: { marginLeft: '0.5rem' } },
                getPageItemProps({ pageValue: totalPages, onPageChange: setPageNumber })),
              [icon('angle-double right', { size: 12 })]
            )
          ]),

          setItemsPerPage && h(Fragment, [
            'Items per page:',
            select({
              style: { marginLeft: '0.5rem' },
              onChange: e => setItemsPerPage(parseInt(e.target.value, 10)),
              value: itemsPerPage
            },
            _.map(i => option({ value: i }, i),
              itemsPerPageOptions))
          ])
        ])
      ]
    )
  ])
}

export const slice = (dataSource, { pageNumber, itemsPerPage }) => {
  return dataSource.slice((pageNumber - 1) * itemsPerPage, pageNumber * itemsPerPage)
}

const defaultComponents = {
  header: {
    row: ({ children, ...props }) => tr(_.merge({
      style: {
        border: '1px solid #CCCCCC',
        backgroundColor: '#FAFAFA'
      }
    }, props), children),
    cell: ({ children, ...props }) => th({
      style: {
        padding: '14px 0',
        fontWeight: 500
      }
    }, [
      div(_.merge({
        style: {
          padding: '0 18px',
          marginRight: -1,
          borderRight: '1px solid #CCCCCC'
        }
      }, props), children)
    ])
  },
  body: {
    row: ({ children, ...props }) => h(Interactive, _.merge({
      as: 'tr',
      style: {
        backgroundColor: 'white',
        border: '1px solid #CCCCCC',
        cursor: null
      },
      hover: { backgroundColor: Style.colors.highlightFaded }
    }, props), children),
    cell: ({ children, ...props }) => td(_.merge({
      style: {
        padding: '16.25px 19px 12.75px',
        overflow: 'hidden'
      }
    }, props), children)
  }
}

export const components = {
  scrollWithHeaderTable: ({
    table: ({ children, ...props }) => table(_.merge({
      style: {
        tableLayout: 'fixed',
        width: '100%'
      }
    }, props), children),
    body: {
      cell: ({ children, ...props }) => td(_.merge({
        style: {
          padding: '16.25px 19px 12.75px',
          whiteSpace: 'nowrap', overflow: 'hidden', textOverflow: 'ellipsis'
        }
      }, props), children)
    }
  }),
  nonInteractiveRow: ({
    body: {
      row: ({ children, ...props }) => tr(_.merge({
        style: {
          backgroundColor: 'white',
          border: '1px solid #CCCCCC'
        }
      }, props), children)
    }
  })
}

/**
 * @param {bool} [allowPagination=true]
 * @param {bool} [allowItemsPerPage=true]
 * @param {number} [defaultItemsPerPage=25]
 * @param {number[]} [itemsPerPageOptions=[10, 25, 50, 100]]
 * @param {function(number)} [onItemsPerPageChanged]
 * @param {number} [initialPage=1]
 * @param {function(number)} [onPageChanged]
 * @param {object[]} dataSource
 * @param {object} tableProps - see {@link https://github.com/react-component/table}, don't provide data
 */
export class DataTable extends Component {
  constructor(props) {
    super(props)
    this.state = {
      pageNumber: props.initialPage || 1,
      itemsPerPage: props.defaultItemsPerPage || 25
    }
  }

  render() {
    const {
      allowPagination = true, allowItemsPerPage = true, itemsPerPageOptions = [10, 25, 50, 100],
      onItemsPerPageChanged, onPageChanged, dataSource, tableProps, customComponents,
      totalRowCount
    } = this.props
    const { pageNumber, itemsPerPage } = this.state

    return h(Fragment, [
      h(RCTable, _.merge({
        data: allowPagination && !totalRowCount ? slice(dataSource, { pageNumber, itemsPerPage }) : dataSource,
        components: _.isArray(customComponents) ?
          _.mergeAll([defaultComponents, ...customComponents]) :
          _.merge(defaultComponents, customComponents)
      }, tableProps)),
      allowPagination ?
        div({ style: { marginTop: 10 } }, [
          paginator({
            filteredDataLength: totalRowCount || dataSource.length,
            setPageNumber: (n => {
              this.setState({ pageNumber: n })
              if (onPageChanged) onPageChanged(n)
            }),
            pageNumber,
            setItemsPerPage: allowItemsPerPage && (n => {
              this.setState({ itemsPerPage: n })
              if (onItemsPerPageChanged) onItemsPerPageChanged(n)
            }),
            itemsPerPage, itemsPerPageOptions
          })
        ]) :
        null
    ])
  }
}

/**
 * @param {bool} [allowPagination=true]
 * @param {bool} [allowItemsPerPage=true]
 * @param {number[]} [itemsPerPageOptions=[12, 24, 36, 48]]
 * @param {function(number)} [onItemsPerPageChanged]
 * @param {function(number)} [onPageChanged]
 * @param {object[]} dataSource
 * @param {function} renderCard - function(record, cardsPerRow) => renderable
 * @param {number} [cardsPerRow=3]
 * @param {number} pageNumber
 * @param {number} itemsPerPage
 */
export class DataGrid extends Component {
  render() {
    const {
      allowPagination = true, allowItemsPerPage = true, itemsPerPageOptions = [12, 24, 36, 48],
      onItemsPerPageChanged, onPageChanged, dataSource, renderCard, cardsPerRow = 3,
      pageNumber, itemsPerPage
    } = this.props

    const listPage = dataSource.slice((pageNumber - 1) * itemsPerPage, pageNumber * itemsPerPage)

    return h(Fragment, [
      div({ style: { display: 'flex', flexWrap: 'wrap' } },
        _.map(record => renderCard(record, cardsPerRow), listPage)),
      allowPagination ?
        div({ style: { marginTop: 10 } }, [
          paginator({
            filteredDataLength: dataSource.length,
            setPageNumber: onPageChanged,
            pageNumber,
            setItemsPerPage: allowItemsPerPage && onItemsPerPageChanged,
            itemsPerPage, itemsPerPageOptions
          })
        ]) :
        null
    ])
  }
}

const cellStyles = {
  display: 'flex',
  alignItems: 'center',
  paddingLeft: '1rem',
  paddingRight: '1rem'
}

const styles = {
  cell: (col, total) => ({
    ...cellStyles,
    borderBottom: `1px solid ${Style.colors.border}`,
    borderLeft: col === 0 ? `1px solid ${Style.colors.border}` : undefined,
    borderRight: col === total - 1 ? `1px solid ${Style.colors.border}` : undefined
  }),
  header: (col, total) => ({
    ...cellStyles,
    backgroundColor: Style.colors.sectionHighlight,
    borderTop: `1px solid ${Style.colors.sectionBorder}`,
    borderBottom: `2px solid ${Style.colors.sectionBorder}`,
    borderLeft: col === 0 ? `1px solid ${Style.colors.sectionBorder}` : undefined,
    borderRight: col === total - 1 ? `1px solid ${Style.colors.sectionBorder}` : undefined,
    borderTopLeftRadius: col === 0 ? '5px' : undefined,
    borderTopRightRadius: col === total - 1 ? '5px' : undefined
  }),
  flexCell: ({ basis = 0, grow = 1, shrink = 1, min = 0, max } = {}) => ({
    flexGrow: grow,
    flexShrink: shrink,
    flexBasis: basis,
    minWidth: min,
    maxWidth: max
  })
}

/**
 * A virtual table with a fixed header and flexible column widths. Intended to take up the full
 * available container width, without horizontal scrolling.
 */

export class FlexTable extends Component {
  constructor(props) {
    super(props)
    this.state = { scrollbarSize: 0 }
  }

  render() {
    const { width, height, rowCount, columns, hoverHighlight } = this.props
    const { scrollbarSize } = this.state
    return div([
      div({
        style: {
          width: width - scrollbarSize,
          height: 48,
          display: 'flex'
        }
      }, [
        ..._.map(([i, { size, headerRenderer }]) => {
          return div({
            key: i,
            style: { ...styles.flexCell(size), ...styles.header(i * 1, columns.length) }
          }, [headerRenderer()])
        }, _.toPairs(columns))
      ]),
      h(RVGrid, {
        width,
        height: height - 48,
        columnWidth: width - scrollbarSize,
        rowHeight: 48,
        rowCount,
        columnCount: 1,
        onScrollbarPresenceChange: ({ vertical, size }) => {
          this.setState({ scrollbarSize: vertical ? size : 0 })
        },
        cellRenderer: data => {
          return h(Interactive, {
            key: data.key,
            as: 'div',
            style: { ...data.style, backgroundColor: '#ffffff', display: 'flex' },
            hover: hoverHighlight ? { backgroundColor: Style.colors.highlightFaded } : undefined
          }, [
            ..._.map(([i, { size, cellRenderer }]) => {
              return div({
                key: i,
                style: { ...styles.flexCell(size), ...styles.cell(i * 1, columns.length) }
              }, [cellRenderer(data)])
            }, _.toPairs(columns))
          ])
        },
        style: { outline: 'none' }
      })
    ])
  }
}

/**
 * A virtual table with a fixed header and explicit column widths. Intended for displaying large
 * datasets which may require horizontal scrolling.
 */

export class GridTable extends Component {
  constructor(props) {
    super(props)
    this.state = { scrollbarSize: 0 }
    this.grid = createRef()
  }

  componentDidMount() {
    this.grid.current.measureAllCells()
  }

  render() {
    const { width, height, rowCount, columns, cellStyle } = this.props
    const { scrollbarSize } = this.state
    return h(RVScrollSync, [
      ({ onScroll, scrollLeft }) => {
        return div([
          h(RVGrid, {
            width: width - scrollbarSize,
            height: 48,
            columnWidth: ({ index }) => columns[index].width,
            rowHeight: 48,
            rowCount: 1,
            columnCount: columns.length,
            cellRenderer: data => {
              return div({
                key: data.key,
                style: { ...data.style, ...styles.header(data.columnIndex, columns.length) }
              }, [
                columns[data.columnIndex].headerRenderer(data)
              ])
            },
            style: { outline: 'none', overflow: 'hidden' },
            scrollLeft,
            onScroll
          }),
          h(RVGrid, {
            ref: this.grid,
            width,
            height: height - 48,
            columnWidth: ({ index }) => columns[index].width,
            rowHeight: 48,
            rowCount,
            columnCount: columns.length,
            onScrollbarPresenceChange: ({ vertical, size }) => {
              this.setState({ scrollbarSize: vertical ? size : 0 })
            },
            cellRenderer: data => {
              return div({
                key: data.key,
                style: {
                  ...data.style,
                  ...styles.cell(data.columnIndex, columns.length),
                  backgroundColor: '#ffffff',
                  ...(cellStyle ? cellStyle(data) : {})
                }
              }, [
                columns[data.columnIndex].cellRenderer(data)
              ])
            },
            style: { outline: 'none' },
            scrollLeft,
            onScroll
          })
        ])
      }
    ])
  }
}

export const TextCell = props => {
  return div(_.merge({
    style: { overflow: 'hidden', whiteSpace: 'nowrap', textOverflow: 'ellipsis' }
  }, props))
}
