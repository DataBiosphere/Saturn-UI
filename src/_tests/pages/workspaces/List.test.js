import { mount } from 'enzyme'
import { h } from 'react-hyperscript-helpers'
import TestRenderer from 'react-test-renderer'
import { DataGrid } from 'src/components/table'
import { waitOneTick, waitOneTickAndUpdate } from 'src/libs/test-utils'
import { WorkspaceList } from 'src/pages/workspaces/List'


describe('WorkspaceList', () => {
  it('should render as expected', () => {
    const renderer = TestRenderer.create(h(WorkspaceList))
    return waitOneTick().then(() => {
      expect(renderer.toJSON()).toMatchSnapshot()
    })
  })

  it('should switch between Grid and List view', () => {
    const wrapper = mount(h(WorkspaceList))
    return waitOneTickAndUpdate(wrapper).then(() => {
      const currentCardsPerRow = () => wrapper.find(DataGrid).props().cardsPerRow

      expect(currentCardsPerRow()).not.toEqual(1)

      wrapper.findIcon('view-list').click()
      expect(currentCardsPerRow()).toEqual(1)

      wrapper.findIcon('view-cards').click()
      expect(currentCardsPerRow()).not.toEqual(1)
    })
  })
})
