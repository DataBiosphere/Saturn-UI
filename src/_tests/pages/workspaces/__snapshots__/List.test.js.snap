// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkspaceList should render as expected 1`] = `
Array [
  <div
    style={
        Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderBottom": "2px solid #2691d0",
            "display": "flex",
            "flex": "none",
            "height": 80,
            "paddingLeft": "1rem",
          }
    }
>
    <clr-icon
        onClick={[Function]}
        shape="bars"
        size={36}
        style={
            Object {
                "color": "#8b5f95",
                "cursor": "pointer",
                "marginRight": "2rem",
              }
        }
    />
    <a
        href={undefined}
        style={
            Object {
                "alignItems": "center",
                "color": "#224f83",
                "display": "flex",
                "fontSize": 22,
                "fontWeight": 500,
                "textTransform": "uppercase",
              }
        }
    >
        <clr-icon
            shape="logoIcon"
            size={48}
            style={
                Object {
                    "marginRight": "0.5rem",
                  }
            }
        />
        <div>
            <div
                style={
                    Object {
                        "color": "#8299a5",
                        "fontSize": "0.8rem",
                        "marginLeft": "0.1rem",
                      }
                }
            >
                Saturn
            </div>
            Workspaces
        </div>
    </a>
    <div
        style={
            Object {
                "borderBottom": "1px solid black",
                "display": "flex",
                "flexGrow": 1,
                "marginLeft": "2rem",
                "maxWidth": 500,
                "padding": "0.5rem 0",
              }
        }
    >
        <clr-icon
            shape="search"
        />
        <input
            onChange={[Function]}
            placeholder="SEARCH BIOSPHERE"
            style={
                Object {
                    "border": "none",
                    "flexGrow": 1,
                    "marginLeft": "1rem",
                    "outline": "none",
                    "verticalAlign": "bottom",
                  }
            }
            value=""
        />
    </div>
</div>,
  <div
    style={
        Object {
            "alignItems": "flex-end",
            "display": "flex",
            "flex": "none",
            "margin": "1rem 4.5rem",
          }
    }
>
    <div
        style={
            Object {
                "color": "#224f83",
                "fontSize": 16,
                "fontWeight": 500,
                "textTransform": "uppercase",
              }
        }
    >
        Workspaces
    </div>
    <div
        style={
            Object {
                "backgroundColor": "white",
                "borderRadius": 3,
                "display": "flex",
                "marginLeft": "auto",
              }
        }
    >
        <div
            disabled={undefined}
            onBlur={[Function]}
            onClick={[Function]}
            onDragEnd={[Function]}
            onDragStart={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onMouseMove={[Function]}
            onMouseUp={[Function]}
            role="button"
            style={
                Object {
                    "alignItems": "center",
                    "color": "#5faee0",
                    "cursor": "pointer",
                    "display": "flex",
                    "height": "2.25rem",
                    "justifyContent": "center",
                    "outline": 0,
                    "outlineOffset": 0,
                    "width": "3rem",
                  }
            }
            tabIndex="0"
        >
            <clr-icon
                shape="view-cards"
                size={24}
            />
        </div>
        <div
            disabled={undefined}
            onBlur={[Function]}
            onClick={[Function]}
            onDragEnd={[Function]}
            onDragStart={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onMouseMove={[Function]}
            onMouseUp={[Function]}
            role="button"
            style={
                Object {
                    "alignItems": "center",
                    "color": "#2691d0",
                    "cursor": "pointer",
                    "display": "flex",
                    "height": "2.25rem",
                    "justifyContent": "center",
                    "outline": 0,
                    "outlineOffset": 0,
                    "width": "3rem",
                  }
            }
            tabIndex="0"
        >
            <clr-icon
                shape="view-list"
                size={24}
            />
        </div>
    </div>
</div>,
  <div
    style={
        Object {
            "display": "flex",
            "flexWrap": "wrap",
            "padding": "0 4rem",
            "position": "relative",
          }
    }
>
    <div
        disabled={undefined}
        onBlur={[Function]}
        onClick={[Function]}
        onDragEnd={[Function]}
        onDragStart={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseMove={[Function]}
        onMouseUp={[Function]}
        role="button"
        style={
            Object {
                "backgroundColor": "white",
                "borderRadius": 5,
                "boxShadow": "0 3px 2px 0 rgba(0,0,0,0.12)",
                "color": "#2691d0",
                "cursor": "pointer",
                "display": "flex",
                "flexDirection": "column",
                "fontSize": 20,
                "height": 225,
                "justifyContent": "center",
                "lineHeight": "28px",
                "margin": "1rem 0.5rem",
                "outline": 0,
                "outlineOffset": 0,
                "padding": "1rem",
                "width": 300,
                "wordWrap": "break-word",
              }
        }
        tabIndex="0"
    >
        <div>
            Create a
        </div>
        <div>
            New Workspace
        </div>
        <clr-icon
            shape="plus-circle"
            size={32}
            style={
                Object {
                    "marginTop": "0.5rem",
                  }
            }
        />
    </div>
    <a
        href={undefined}
        style={
            Object {
                "backgroundColor": "white",
                "borderRadius": 5,
                "boxShadow": "0 3px 2px 0 rgba(0,0,0,0.12)",
                "display": "flex",
                "flexDirection": "column",
                "height": 225,
                "justifyContent": "space-between",
                "margin": "1rem 0.5rem",
                "padding": "1rem",
                "width": 300,
                "wordWrap": "break-word",
              }
        }
    >
        <div
            style={
                Object {
                    "color": "#2691d0",
                    "flex": "none",
                    "fontSize": 16,
                    "height": "40px",
                    "lineHeight": "20px",
                    "overflow": "hidden",
                    "wordWrap": "break-word",
                  }
            }
        >
            large_sample_copy
        </div>
        <div
            style={
                Object {
                    "flex": "none",
                    "height": "90px",
                    "lineHeight": "18px",
                    "overflow": "hidden",
                  }
            }
        >
            <span
                style={
                    Object {
                        "color": "#a6a6a6",
                      }
                }
            >
                No description added
            </span>
        </div>
        <div
            style={
                Object {
                    "alignItems": "center",
                    "display": "flex",
                  }
            }
        >
            <div
                style={
                    Object {
                        "flex": 1,
                      }
                }
            >
                Last changed: ***MOCKED DATE***
            </div>
            <div
                style={
                    Object {
                        "flex": "none",
                      }
                }
            >
                <div
                    style={
                        Object {
                            "backgroundColor": "#8b5f95",
                            "borderRadius": "1.5rem",
                            "color": "white",
                            "height": "1.5rem",
                            "lineHeight": "1.5rem",
                            "textAlign": "center",
                            "width": "1.5rem",
                          }
                    }
                    title="alice@example.com"
                >
                    A
                </div>
            </div>
        </div>
    </a>
    <a
        href={undefined}
        style={
            Object {
                "backgroundColor": "white",
                "borderRadius": 5,
                "boxShadow": "0 3px 2px 0 rgba(0,0,0,0.12)",
                "display": "flex",
                "flexDirection": "column",
                "height": 225,
                "justifyContent": "space-between",
                "margin": "1rem 0.5rem",
                "padding": "1rem",
                "width": 300,
                "wordWrap": "break-word",
              }
        }
    >
        <div
            style={
                Object {
                    "color": "#2691d0",
                    "flex": "none",
                    "fontSize": 16,
                    "height": "40px",
                    "lineHeight": "20px",
                    "overflow": "hidden",
                    "wordWrap": "break-word",
                  }
            }
        >
            large_sample_copy
        </div>
        <div
            style={
                Object {
                    "flex": "none",
                    "height": "90px",
                    "lineHeight": "18px",
                    "overflow": "hidden",
                  }
            }
        >
            <span
                style={
                    Object {
                        "color": "#a6a6a6",
                      }
                }
            >
                No description added
            </span>
        </div>
        <div
            style={
                Object {
                    "alignItems": "center",
                    "display": "flex",
                  }
            }
        >
            <div
                style={
                    Object {
                        "flex": 1,
                      }
                }
            >
                Last changed: ***MOCKED DATE***
            </div>
            <div
                style={
                    Object {
                        "flex": "none",
                      }
                }
            >
                <div
                    style={
                        Object {
                            "backgroundColor": "#8b5f95",
                            "borderRadius": "1.5rem",
                            "color": "white",
                            "height": "1.5rem",
                            "lineHeight": "1.5rem",
                            "textAlign": "center",
                            "width": "1.5rem",
                          }
                    }
                    title="alice@example.com"
                >
                    A
                </div>
            </div>
        </div>
    </a>
</div>,
]
`;
